mu/app.py

```diff
@@ -79,6 +93,40 @@ def set_frame(self):
         self.setPixmap(pixmap)
         self.setMask(pixmap.mask())
 
+    def draw_log(self, text):
+        """
+        Draw the log entries onto the splash screen. Will only display the last
+        self.log_lines number of log entries. The logs will be displayed at the
+        bottom of the splash screen, justified left.
+        """
+        self.log.append(text)
+        self.log = self.log[-self.log_lines :]
+        if self.log:
+            self.draw_text("\n".join(self.log))
+
+    def draw_text(self, text):
+        """
+        Draw text into splash screen.
+        """
+        if text:
+            self.showMessage(text, Qt.AlignBottom | Qt.AlignLeft)
+
+    def failed(self, text):
+        """
+        Something has gone wrong during start-up, so signal this, display a
+        helpful message along with instructions for what to do.
+        """
+        self.animation.stop()
+        pixmap = load_pixmap("splash_fail.png")
+        self.setPixmap(pixmap)
+        lines = text.split("\n")
+        lines.append(
+            "This screen will close in a few seconds. "
+            + "Then a crash report tool will open in your browser."
+        )
+        lines = lines[-12:]
```

Comment:
Not sure where 12 comes from? if `text` is the exception string, shoudn't we print the entire thing?
Is this related to the size of the splash screen?

Reply:
This is related to the size of the splash screen. When the crash report runs, the **whole** exception string is used.

I limit this because we simply don't know how big the stack trace is going to be, and I want Mu to appear as if it's handling things gracefully. No information is lost, and the user always see the bottom of the error output (which is usually what really caused the problem).